### ISSUE TITLE
🌮 Proposal: Polymorphic `as` prop to manipulate rendered HTML tag/element.

### ISSUE THREAD

**Issue by codebender828** [styled-components/vue-styled-components#92]
# 🌮 Proposal: Polymorphic `as` prop to adapt tag

Thanks to the Vue styled components team for the work done with this project. I'm currently using it to build a components library and I love it! There is a challenge though that I have afore mentioned in the title to this issue.

### 📚 The challenge with `withComponent` API?
Despite the `.withComponent` API is sufficient to change the tags for styled component of single depth, it still remains fairly cumbersome and verbose to manipulate the HTML tag/element with which a styled component is rendered without calling the `.withComponent` API.

Furthermore, a `StyledComponent` derived from the `.withComponent` does not inherit style props from low-level styled component. i.e. If the parent component(`PseudoBox`) is composed from another base styled component(`Box`), the resulting component(`Button`) will not inherit style props from the grandparent styled component(`Box`). 

I made this [CodeSandbox](https://codesandbox.io/s/why-withcomponent-needs-to-be-supplemented-with-astag-polymorphic-prop-kh32i) to illustrate this problem.
https://codesandbox.io/s/why-withcomponent-needs-to-be-supplemented-with-astag-polymorphic-prop-kh32i

This is not a new problem. React's styled-components' `.withComponent` equivalent [also faces a similar problem.](https://www.styled-components.com/docs/api#withcomponent)

>As of styled-components v4 the withComponent API is now a candidate for deprecation. In all likelihood, you probably want to use the new "as" prop to simply switch what element/component being rendered since the withComponent API is destructive toward styles if the lowest-wrapped component is a StyledComponent.

## 🌮 Why a polymorphic `as`/`tag` prop?
What the polymorphic `as` prop does is essentially allow `vue-styled-components` to declaratively and dynamically change the HTML tag with which a StyledComponent is rendered. Similar to React's [polymorphic `as`](https://www.styled-components.com/docs/api#as-polymorphic-prop), it would allow you to keep all the styling you've applied to a `StyledComponent` but just switch out what's being ultimately rendered (be it a different HTML tag or a different custom component).

**Example**
In Component.js
```js
//  Renders a div element by default.
const Component = styled('div', {})``
```
Using the `as` prop in another template/component would be as shown below.
```vue
<template>
  <!-- Component will render a button instead of a div -->
  <!-- This will also preserve style props from low-level StyledComponents -->
  <Component as="button" color="red">
    Button
  </Component>
</template>
```

**Comment by codebender828** [comment]
I would like to be able to contribute towards implementing something like this to supplement the `.withComponent` API.

**Comment by liqueflies** [comment]
Merged #93 and added in **1.5.0** :)
Thank you <3
